Userland/Games/TypingTutor/TypingTutorCanvasWidget.cpp

```diff
@@ -0,0 +1,227 @@
+    }
+
+    file->close();
+    m_wordlist_loaded = true;
+}
+
+void TypingTutorCanvasWidget::generate_new_wave()
+{
+    u8 glyph_height = font().glyph_height();
+    u8 max_wave_size = s_height / glyph_height - 1;
+
+    for (size_t i = 0; i < m_current_wave_size; ++i) {
+        Word word = {
+            .x = (float)random_between(0, 50),
+            .y = random_between(1, max_wave_size),
+            .velocity = (float)random_between(3 * m_max_velocity / 4, m_max_velocity) / s_frame_rate,
+            .value = m_wordlist[random_between(0, m_wordlist.size() - 1)],
+        };
+
+        for (size_t j = 0; j < m_wave.size(); ++j) {
+            while (word.value == m_wave[j].value)
+                word.value = m_wordlist[random_between(0, m_wordlist.size() - 1)];
+
+            if (word.y == m_wave[j].y) {
+                size_t word_width = font().width(word.value);
+                if (word.x + word_width >= m_wave[j].x) {
+                    float random_offset = word.x;
+                    word.x = m_wave[j].x - word_width;
+                    word.x -= random_offset + 10;
+                }
+
+                if (word.velocity > m_wave[j].velocity)
+                    word.velocity = m_wave[j].velocity;
+            }
+        }
+
+        m_wave.append(word);
+    }
+}
+
+bool TypingTutorCanvasWidget::delete_word_if_correct(String attempt)
+{
+    for (int i = m_wave.size() - 1; i >= 0; --i) {
+        if (m_wave[i].value == attempt) {
+            m_total_characters_wrote += attempt.length();
+            m_wave.remove(i);
+            return true;
+        }
+    }
+
+    return false;
+}
+
+float TypingTutorCanvasWidget::speed()
+{
+    float t = elapsed_time() / 60;
+    if (t == 0)
+        return 0;
+
+    return m_total_characters_wrote / t;
+}
+
+float TypingTutorCanvasWidget::elapsed_time()
+{
+    return (float)m_frame_count / s_frame_rate;
+}
+
+void TypingTutorCanvasWidget::paint_event(GUI::PaintEvent& event)
+{
+    GUI::Widget::paint_event(event);
+
+    GUI::Painter painter(*this);
+    painter.add_clip_rect(event.rect());
+    painter.fill_rect(event.rect(), Color::Black);
+
+    String wave_info = String::formatted("[ Next wave in: {} s ]", m_next_wave_countdown);
+    int text_width = font().width(wave_info);
+    u8 glyph_height = font().glyph_height();
+    Gfx::IntRect wave_info_rect((int)s_width / 2 - text_width / 2, 0, text_width, glyph_height);
+    painter.draw_text(wave_info_rect, wave_info, Gfx::TextAlignment::TopLeft, Color::White);
+    painter.draw_line({ 0, (int)glyph_height / 2 }, { (int)s_width / 2 - text_width / 2, glyph_height / 2 }, Color::White);
+    painter.draw_line({ (int)s_width / 2 + (int)text_width / 2, (int)glyph_height / 2 }, { s_width, glyph_height / 2 }, Color::White);
+
+    painter.fill_rect({ 0, glyph_height, s_width, s_height }, Color::from_rgb(0x2631ad));
+
+    for (size_t i = 0; i < m_wave.size(); ++i)
+        draw_word(m_wave[i], painter);
```

Comment:
Consider using range based for loops instead of manually iterating?

```suggestion
     for (auto& wave: m_wave)
        draw_word(wave, painter);
```

Reply:
Yeah, sure I'll do it.